### PR TITLE
Feature/add gcov target

### PR DESCRIPTION
**Pull request by ClausKlein**
Adds test coverage report feature. 
Can be invoked with `cmake process_coverage`, the coverage report is generated in `build/coverage/coverage.html`.
Needs client to install `gcovr`.

**NOTE**: This is **second** of a series of PR based on https://github.com/bemanproject/inplace_vector/pull/14

- [ ] **Depends** on https://github.com/bemanproject/inplace_vector/pull/33

see too https://github.com/bemanproject/optional26/pull/82

### CODE REVIEW

**Review by wusatosi — CHANGES_REQUESTED** [review]
Thank you for splitting this change out as a separate PR.

Can you please keep the domain of a PR restricted to add gcov target.
No need to rush here.

Please include a description of the coverage feature in the PR, instead of:
```
Added VERIFY_INTERFACE_HEADER_SETS too.
(and fixed compiler errors found with this check)

NOTE: This is first of a series of PR based on https://github.com/bemanproject/inplace_vector/pull/14
```

Change to, e.g.
```
Adds test coverage report feature. 
Can be invoked with `cmake process_coverage`, the coverage report is generated in `build/coverage/coverage.html`.
Needs client to install `gcovr`.

Note: This is first of a series of PR based on #14 .
```

Request change:
- remove unrelated CL to adding gcov target
- Include a short description of this cmake target in README
- Add license header for `cmake/gcovr.cfg.in`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,41 +33,64 @@ option(
     ${PROJECT_IS_TOP_LEVEL}
 )
 
+set(CPACK_GENERATOR TGZ)
+
+# includes
 include(GNUInstallDirs)
+include(CPack)
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:40]: done

File: CMakeLists.txt [hunk 2]
```diff
@@ -3,7 +3,9 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 # cmake-format: on
 
-cmake_minimum_required(VERSION 3.23)
+set(CMAKE_SKIP_TEST_ALL_DEPENDENCY FALSE)
+
+cmake_minimum_required(VERSION 3.25...3.31)
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:8]: done

File: CMakeLists.txt [hunk 3]
```diff
@@ -27,41 +33,64 @@ option(
...
     INTERFACE
-        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-        $<INSTALL_INTERFACE:include>
+        "$<$<COMPILE_FEATURES:cxx_std_23>:cxx_std_23>"
+        "$<$<NOT:$<COMPILE_FEATURES:cxx_std_23>>:cxx_std_20>"
 )
 
+set(TARGET_PACKAGE_NAME ${PROJECT_NAME}-config)
+set(TARGETS_EXPORT_NAME ${PROJECT_NAME}-targets)
+set(INSTALL_CONFIGDIR ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME})
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:68]: done

File: CMakeLists.txt [hunk 4]
```diff
@@ -27,41 +33,64 @@ option(
...
 
 if(BEMAN_EXEMPLAR_BUILD_EXAMPLES)
     add_subdirectory(examples)
 endif()
+
+# Coverage
+configure_file(cmake/gcovr.cfg.in gcovr.cfg @ONLY)
+
+add_custom_target(
+    process_coverage
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:90]: yes

File: cmake/gcovr.cfg.in [hunk 5]
```diff
@@ -0,0 +1,11 @@
+root = @CMAKE_SOURCE_DIR@
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply cmake/gcovr.cfg.in:1]: yes

File: .github/workflows/ci_tests.yml [hunk 6]
```diff
@@ -90,15 +90,15 @@ jobs:
...
+          cmake --build build --config Release --target all_verify_interface_header_sets
           cmake --install build --config Release --prefix /tmp/beman.inplace_vector
           find /tmp/beman.inplace_vector -type f
       - name: Test Release
         run: ctest --test-dir build --build-config Release
       - name: Build Debug
         run: |
           cmake --build build --config Debug --verbose
-          # cmake --build build --config Debug --target all_verify_interface_header_sets
+          cmake --build build --config Debug --target all_verify_interface_header_sets
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply .github/workflows/ci_tests.yml:101]: done

File: CMakeLists.txt [hunk 7]
```diff
@@ -27,41 +33,64 @@ option(
...
-)
-
-# Install the header files to the appropriate destination
-install(
-    DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/
-    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${CMAKE_PROJECT_NAME}
-    FILES_MATCHING
-    PATTERN
-    "${CMAKE_CURRENT_SOURCE_DIR}/include/beman/inplace_vector/inplace_vector.hpp"
+    FILE_SET inplace_vector_public_headers
```
[the remark this answers is not in the document]
Comment (ClausKlein) [reply CMakeLists.txt:74]: done

File: CMakeLists.txt [hunk 8]
```diff
@@ -13,6 +15,10 @@ project(
     LANGUAGES CXX
 )
 
+if(${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+    message(FATAL_ERROR "In-source builds are not allowed!")
+endif()
```
[the remark this answers is not in the document]
Comment (wusatosi) [reply CMakeLists.txt:20]: This is not done. Please exclude.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (wusatosi) [reply CMakeLists.txt:40]: Please exclude.

File: CMakeLists.txt [hunk 9]
```diff
@@ -27,41 +33,64 @@ option(
...
+        TYPE HEADERS
+        BASE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/include
+        FILES
+            "${CMAKE_CURRENT_SOURCE_DIR}/include/beman/inplace_vector/inplace_vector.hpp"
+)
+set_target_properties(
+    beman.inplace_vector
+    PROPERTIES VERIFY_INTERFACE_HEADER_SETS ON
+)
+target_compile_features(
```
Comment (wusatosi) [inline CMakeLists.txt:59]: Please exclude unrelated changes

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (wusatosi) [reply CMakeLists.txt:68]: Please exclude

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (wusatosi) [reply CMakeLists.txt:74]: Please exclude

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (wusatosi) [reply CMakeLists.txt:90]: Can you add a CI job to test this.

File: cmake/gcovr.cfg.in [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception.
+
+root = @CMAKE_SOURCE_DIR@
+cobertura = @CMAKE_BINARY_DIR@/coverage/cobertura.xml
+sonarqube = @CMAKE_BINARY_DIR@/coverage/sonarqube.xml
+html-details = @CMAKE_BINARY_DIR@/coverage/coverage.html
+# XXX gcov-executable = @GCOV_EXECUTABLE@
```
Comment (wusatosi) [inline cmake/gcovr.cfg.in:7]: What is this comment for

**Review by wusatosi — CHANGES_REQUESTED** [review]
Please rebase and address previous concerns.

File: cmake/gcovr.cfg.in (hunk 10 above)
Comment (ClausKlein) [reply cmake/gcovr.cfg.in:7]: This is based on an other beman project, but this use presets ...
